src/tests/stress/pedsim/Ped_util/Quad_tree.enc

```diff
@@ -0,0 +1,154 @@
+    this.max_x = max_x
+    this.max_y = max_y
+    this.min_x = min_x
+    this.min_y = min_y
+    this.top_right=Nothing
+    this.top_left=Nothing
+    this.bottom_right=Nothing
+    this.bottom_left=Nothing
+    this.size = 0
+    this.agent = new [int](0)
+  end
+  def add(x : int, y : int) : unit
+    --if x <= this.max_x && x >= this.min_x && y <= this.max_y && y >= this.min_y then
+      match this.top_right with
+          case Nothing =>
+            if |this.agent|==0 then
+                this.agent = new [int](2)
+                this.agent(0) = x
+                this.agent(1) = y
+                this.size = 1
+            else
+              val dx = (this.max_x - this.min_x) / 2
+              val dy = (this.max_y - this.min_y) / 2
+              this.top_right = Just(new Quad(this.max_x, this.max_y, this.min_x + dx + 1, this.min_y + dy + 1))
+              this.top_left = Just(new Quad(this.min_x + dx, this.max_y, this.min_x, this.min_y + dy + 1))
+              this.bottom_right = Just(new Quad(this.max_x, this.min_y + dy, this.min_x + dx + 1, this.min_y))
+              this.bottom_left = Just(new Quad(this.min_x + dx, this.min_y + dy, this.min_x, this.min_y))
+              this.size = 2
+              --print("deeper {} {} {} {} {} {}\n", this.max_x, this.max_y, x, y,this.agent(0),this.agent(1))
+              this.recur(this.agent(0), this.agent(1)).add(this.agent(0), this.agent(1))
+              this.recur(x, y).add(x, y)
+              this.agent = new [int](0)
+            end
+          end
+          case unused =>
+              this.size = this.size + 1
+              this.recur(x, y).add(x, y)
+          end
+      end
+  end
+  def remove(x : int, y : int) : bool
+    match this.top_right with
+      case Nothing =>
+          if this.same(x, y) then
+            this.agent = new[int](0)
+            this.size = 0
+            true
+          else
+            false
+          end
+      end
+      case Just(unused) =>
+          val hit = this.recur(x, y).remove(x, y)
+          if hit then
+            this.size = this.size - 1
+            if this.size == 0 then
+              this.top_right = Nothing
+              this.bottom_right = Nothing
+              this.top_left = Nothing
+              this.bottom_left = Nothing
+            end
+          end
+          hit
+      end
+    end
+  end
+  def isin(x : int, y : int) : bool
+    match this.top_right with
+      case Nothing => this.same(x, y)
+      case Just(unused) => this.recur(x, y).isin(x, y)
+    end
+  end
+  def same(x : int, y : int) : bool
+    if |this.agent| == 2 && this.agent(0) == x && this.agent(1) == y then
+      true
+    else
+      false
+    end
+  end
+  def recur(x : int, y : int) : Quad
+    val top = y > (((this.max_y - this.min_y) / 2) + this.min_y)
+    val right = x > (((this.max_x - this.min_x) / 2) + this.min_x)
+    if top then
+      if right then
+        match (this.top_right) with
+            case Just(qq) => qq
```

Comment:
You can't, recur only gets called after the node is known to not be a leaf. If you did tho, you would crash, which is what we want in that case.